### ISSUE TITLE
Unexposed input of hostDirective hide input of directive on host element

### ISSUE THREAD

**Issue by GuillaumeNury** [angular/angular#52072]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

We we use a directive in `hostDirectives` and on host element, two instances are created:

```ts
@Directive({
	selector: '[appTest]',
	standalone: true,
})
class TestDirective {
	constructor() {
		console.log('TestDirective'); // logs two times
	}
}

@Component({
	selector: 'app-test',
	template: 'test',
	standalone: true,
	hostDirectives: [TestDirective], // <-- here
})
class TestComponent {}

@Component({
	selector: 'app-host',
	template: '<app-test appTest></app-test>',
        //                   ^^^^^^^ and here
	standalone: true,
	imports: [TestComponent, TestDirective],
})
class HostComponent {}
```

But if I add an input to the previous directive, I can no longer use it in `HostElement`. Event if `TestComponent` does not expose `TestDirective`'s inputs:

```ts
@Directive({
	selector: '[appTest]',
	standalone: true,
})
class TestDirective implements OnInit {
	@Input() appTest: string;

	ngOnInit() {
		console.log('TestDirective', this.appTest); // logs two times, once with 321 and once with undefined
	}
}

@Component({
	selector: 'app-test',
	template: 'test',
	standalone: true,
	hostDirectives: [TestDirective],
})
class TestComponent {
	constructor() {
		inject(TestDirective).appTest = 321;
	}
}

@Component({
	selector: 'app-host',
	template: '<app-test [appTest]="123"></app-test>',
       //                    ^^^^^^^^^ NG0303: Can't bind to 'appTest' since it isn't a known property of 'app-test'
	standalone: true,
	imports: [TestComponent, TestDirective],
})
class HostComponent {}
```

A more realistic example is with `NgClass`. When you have many classes with dynamic names, it is easier to use a `NgClass` hostDirective than to use a `HostBinding('class')`. But doing so prevent the consumer of the component to use `ngClass` on it.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-sezuuc?file=src%2Fmain.ts,src%2Fwith-input.ts,src%2Fwithout-input.ts

### Please provide the exception or error you saw

```true
core.mjs:6203 NG0303: Can't bind to 'appTest' since it isn't a known property of 'app-test' (used in the 'HostComponent' component template).
1. If 'app-test' is an Angular component and it has the 'appTest' input, then verify that it is included in the '@Component.imports' of this component.
2. If 'app-test' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@Component.schemas' of this component to suppress this message.
3. To allow any property add 'NO_ERRORS_SCHEMA' to the '@Component.schemas' of this component.
```
### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.2.5
Node: 18.12.1
Package Manager: npm 8.19.2
OS: linux x64

Angular: 16.2.8
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1602.5
@angular-devkit/build-angular   16.2.5
@angular-devkit/core            16.2.5
@angular-devkit/schematics      16.2.5
@angular/cdk                    16.2.7
@angular/cli                    16.2.5
@schematics/angular             16.2.5
ng-packagr                      16.0.0
rxjs                            7.8.1
typescript                      5.0.4
zone.js                         0.13.0
```

### Anything else?

In the meantime, we can "copy" NgClass. But it is less elegant:

```ts
@Directive({
	selector: '[ngClazz]',
	standalone: true,
})
class NgClazz extends NgClass {} // Use this class in `hostDirectives`
```

**Comment by crisbeto** [comment]
You have to explicitly expose the inputs from the host directive, e.g. `hostDirectives: [{directive: TestDirective, inputs: ['appTest']}]`.

**Comment by crisbeto** [comment]
I looked into it a bit and I think it's a problem in our validation. We have a check to prevent the same directive from being applied twice to the same element, but it only worked with directives, not with components.

**Comment by crisbeto** [comment]
I've sent out a fix at #52073, but it's worth noting that it only fixes our validation. What the example above is trying to do isn't allowed, because a directive isn't allowed to match on an element more than once.

**Comment by GuillaumeNury** [comment]
It makes sense to forbid it. Both directive instance could not know each other. And the Host component could not inject the "outside" instance.

The work around of "cloning" the directive is not that messy and solve the problem above.

Thanks for the lightning fast response and fix. Really appreciate Angular team's work !
